### PR TITLE
MAINT: Replace usage of np.prod

### PR DESCRIPTION
**Pull request by eendebakpt**
We replace usage of `np.prod` with either a direct call to `array.size` if possible or else use `math.prod`.
Direct usage of the size leads to cleaner code. Both alternatives are significantly faster. For example:
```    
x = [1, 2, 3]

%timeit math.prod(x)
50.7 ns ± 2.24 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

%timeit np.prod(x)
2.53 µs ± 36 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

@ev-br This is similar to #20325. These are all the easy/clear cases I could find.

### CODE REVIEW

File: scipy/io/matlab/_mio5.py [hunk 1]
```diff
@@ -68,7 +68,7 @@
 Henkelmann; parts of the code for simplify_cells=True adapted from
 http://blog.nephics.com/2019/08/28/better-loadmat-for-scipy/.
 '''
-
+import math
```
Comment (lucascolley) [inline scipy/io/matlab/_mio5.py:71]:
```suggestion

import math
```

File: scipy/sparse/_data.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
     as self but with a different data array
 
 """
-
+import math
```
Comment (lucascolley) [inline scipy/sparse/_data.py:8]:
```suggestion

import math
```

**Review by lucascolley — CHANGES_REQUESTED** [review]

**Comment by lucascolley** [comment]
remaining CI failure unrelated

**Comment by ev-br** [comment]
LGTM.
Review comments seem to be all addressed, the CI failure is unrelated and has been fixed in main already. Let's land this. Thanks @eendebakpt , all.

**Pull request merged**